pkg/operator/image.go

```diff
@@ -17,26 +17,34 @@ package operator
 import (
 	"fmt"
 	"strings"
+
+	dockerref "github.com/docker/distribution/reference"
 )
 
 // BuildImagePath builds a container image path based on
 // the given parameters.
-// baseImage and version are used by default.
-// If the tag is specified, we use the tag to identify the container image.
-// If the sha is specified, we use the sha to identify the container image,
-// as it has even stronger immutable guarantees to identify the image.
-func BuildImagePath(baseImage, version, tag, sha string) string {
-	image := baseImage
-	if version != "" {
-		image = fmt.Sprintf("%s:%s", baseImage, version)
-	}
-	if tag != "" {
-		image = fmt.Sprintf("%s:%s", baseImage, tag)
+func BuildImagePath(image, version, tag, sha string) (string, error) {
+	// If the image contains no tag or digest, a default tag should be applied.
+	named, err := dockerref.ParseNormalizedNamed(image)
+	if err != nil {
+		return "", fmt.Errorf("couldn't parse image reference %q: %v", image, err)
 	}
-	if sha != "" {
-		image = fmt.Sprintf("%s@sha256:%s", baseImage, sha)
+	_, isTagged := named.(dockerref.Tagged)
+	_, isDigested := named.(dockerref.Digested)
+	if !isTagged && !isDigested {
```

Comment:
nit: instead of having a big if block here, return early:

```
if isTagged || isDigested {
  return image, nil
}
```
and reindent the complex if-else logic:
```
  if sha != " { ...
  } else if version != "" {
 ..
  }
return image, nil
```